components/requestreply/command_bus.go

```diff
@@ -0,0 +1,80 @@
+package requestreply
+
+import (
+	"context"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+type CommandBus interface {
+	SendWithModifiedMessage(ctx context.Context, cmd any, modify func(*message.Message) error) error
+}
+
+// SendWithReply sends command to the command bus and receives a replies of the command handler.
+// It returns a channel with replies, cancel function and error.
+//
+// SendWithReply can be cancelled by calling cancel function or by cancelling context or
+// When SendWithReply is canceled, the returned channel is closed as well.
+// by exceeding the timeout set in the backend (if set).
+// Warning: It's important to cancel the function, because it's listening for the replies in the background.
+// Lack of cancelling the function can lead to subscriber leak.
+//
+// SendWithReply can listen for handlers with results (NewCommandHandlerWithResult) and without results (NewCommandHandler).
+// If you are listening for handlers without results, you should pass `NoResult` or `struct{}` as `Result` generic type:
+//
+//	 replyCh, cancel, err := requestreply.SendWithReply[requestreply.NoResult](
+//			context.Background(),
+//			ts.CommandBus,
+//			ts.RequestReplyBackend,
+//			&TestCommand{ID: "1"},
+//		)
+//
+// If `NewCommandHandlerWithResult` handler returns a specific type, you should pass it as `Result` generic type:
+//
+//	 replyCh, cancel, err := requestreply.SendWithReply[SomeTypeReturnedByHandler](
+//			context.Background(),
+//			ts.CommandBus,
+//			ts.RequestReplyBackend,
+//			&TestCommand{ID: "1"},
+//		)
+//
+// SendWithReply will send the replies to the channel until the context is cancelled or the timeout is exceeded.
+// They are multiple cases when more than one reply can be sent:
+//   - when the handler returns an error, and backend is configured to nack the message on error
+//     (for the PubSubBackend, it depends on `PubSubBackendConfig.AckCommandErrors` option.),
+//   - when you are using fan-out mechanism and commands are handled multiple times,
```

Comment:
```suggestion
// SendWithReply sends a command to the command bus and receives replies from the command handler(s).
// It returns a channel with replies, cancel function, and error.
//
// SendWithReply can be canceled by calling the cancel function, canceling the context, or exceeding the timeout set in the backend (if set).
// When SendWithReply is canceled, the returned channel is closed as well.
// 
// Warning: It's important to cancel the function because it listens for the replies in the background.
// Not canceling it can lead to subscriber leaks.
//
// SendWithReply can listen for handlers with results (NewCommandHandlerWithResult) and without results (NewCommandHandler).
// When listening for handlers without results, you should pass `NoResult` or `struct{}` as the `Result` generic type:
//
//	 replyCh, cancel, err := requestreply.SendWithReply[requestreply.NoResult](
//			context.Background(),
//			ts.CommandBus,
//			ts.RequestReplyBackend,
//			&TestCommand{ID: "1"},
//		)
//
// If the `NewCommandHandlerWithResult` handler returns a specific type, you should pass it as the `Result` generic type:
//
//	 replyCh, cancel, err := requestreply.SendWithReply[SomeTypeReturnedByHandler](
//			context.Background(),
//			ts.CommandBus,
//			ts.RequestReplyBackend,
//			&TestCommand{ID: "1"},
//		)
//
// SendWithReply will send the replies to the channel until the context is canceled or the timeout is exceeded.
//There are multiple cases when more than one reply can be sent:
//   - when the handler returns an error, and the backend is configured to nack the message on error
//     (for the PubSubBackend, it depends on the `PubSubBackendConfig.AckCommandErrors` option),
//   - when you use a fan-out mechanism and commands are handled multiple times by more than one handler).
```